include/miral/miral/idle_listener.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIRAL_IDLE_LISTENER_H
+#define MIRAL_IDLE_LISTENER_H
+
+#include <functional>
+#include <memory>
+
+namespace mir { class Server; }
+
+namespace miral
+{
+
+
+class IdleListener
+{
+public:
+    using Callback = std::function<void()>;
+    struct Callbacks
+    {
+        Callback on_dim = []{};
+        Callback on_off = []{};
+        Callback on_wake = []{};
+    };
+
+    explicit IdleListener(Callbacks const&);
```

Comment:
I have been hasty or imprecise in my comments a few iterations back. I see no improvement from:

```c++
IdleListener foo{[&]{stuff}, [&]{more-stuff}, [&]{even-more-stuff}};
```
to what we have now:

```c++
IdleListener foo{{[&]{stuff}, [&]{more-stuff}, [&]{even-more-stuff}}};
```
I.e. multiple parameters of the same type in some order.

I think this would be better:

```c++
auto foo = IdleListener{}
    .on_dim([&]{stuff})
    .on_off([&]{more-stuff}
    .on_wake([&]{even-more-stuff});
```